compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/collectors/components/ErrorNodeDiagnosticCollectorComponent.kt

```diff
@@ -54,15 +57,55 @@ class ErrorNodeDiagnosticCollectorComponent(collector: AbstractDiagnosticCollect
         ) return
 
         // If the receiver cannot be resolved, we skip reporting any further problems for this call.
-        if (qualifiedAccess?.dispatchReceiver.hasUnresolvedNameError() || qualifiedAccess?.extensionReceiver.hasUnresolvedNameError() || qualifiedAccess?.explicitReceiver.hasUnresolvedNameError()) return
+        if (qualifiedAccess?.dispatchReceiver.hasUnresolvedNameError() ||
+            qualifiedAccess?.extensionReceiver.hasUnresolvedNameError() ||
+            qualifiedAccess?.explicitReceiver.hasUnresolvedNameError()
+        ) return
+
+        if (reportUninitializedParameter(qualifiedAccess, data)) {
+            return
+        }
 
         reportFirDiagnostic(errorNamedReference.diagnostic, source, reporter, data, qualifiedAccess?.source)
     }
 
     private fun FirExpression?.hasUnresolvedNameError(): Boolean {
-        return when ((this?.typeRef as? FirErrorTypeRef)?.diagnostic) {
-            is ConeUnresolvedNameError -> true
-            else -> false
+        return this?.unresolvedNameError != null
+    }
+
+    private val FirExpression.unresolvedNameError: ConeUnresolvedNameError?
+        get() {
+            val typeRef = if (this is FirAnonymousFunction) this.returnTypeRef else this.typeRef
+            return (typeRef as? FirErrorTypeRef)?.diagnostic as? ConeUnresolvedNameError
+        }
+
+    private fun reportUninitializedParameter(qualifiedAccess: FirQualifiedAccessExpression?, context: CheckerContext): Boolean {
+        if (qualifiedAccess == null) return false
+        val unresolvedNameError = qualifiedAccess.unresolvedNameError ?: return false
+
+        val valueParameter = context.containingDeclarations.lastOrNull { valueParameter ->
+            valueParameter is FirValueParameter &&
+                    valueParameter.defaultValue?.unresolvedNameError == unresolvedNameError
```

Comment:
Checking `defaultValue` won't always work, example:
```kt
fun foo(x: () -> Int = { y; 1 }, y: Int) {}
```
The default value of `x` is not an error since the unresolved `y` is not the last statement in the lambda. The default value type is `Int`. Even a lambda like `{ if (true) y else 1 }` (default value type is still `Int`, but not if the branches are reversed).

Do you _need_ to check the default value? Isn't the containing value parameter _always_ going to be the correct value parameter, regardless of default value? Actually, no, because if you want a really ridiculous test, try:
```kt
fun foo(x: () -> Int = {
    fun bar(xx: () -> Int = { y; 1 }) = xx
    bar()()
}, y: Int) {}
```
So you can't rely on either the default value or the closest containing value parameter (it is nested deep).

Here is another case to consider:
```kt
fun foo(x: () -> Int = {
    fun bar(): Int = y
    bar()
}, y: Int) {}
```

...this is tricky because the closest containing value parameter (`x`) does not belong to the closest containing function (`bar`) to the qualified access.

One solution I can think of is start with the closest containing value parameter, go back up one declaration to get the function, and then find the referred parameter. If not found, go to the next closest containing value parameter and repeat. I think this is actually simpler and in 99.9999% of cases, there would be no need to go up multiple value parameters.

Reply:
Thank you for interesting corner cases, you rock!

I didn't worry about the 3rd case, since the function lookup below checks if the current function's value parameter has a referred parameter name or not. So, I believe it would correctly pick `foo` instead of `bar`.

Rather, my concern was 2nd case, which really needs a loop that walks upward until the pair of value parameter _and_ function were found. Then, I realized all of this ugly context traversal is attributable to unwanted, unresolved error for value parameter reference. That led me to a complete rework: #4321.